### PR TITLE
Reconcile lockbox 27

### LINKED ISSUE #27: Clinic user can reconcile lockbox
As a clinic user, I want to log the amount of money in the lockbox once per month and make any necessary adjustments.

image

image

### PR DESCRIPTION
**Pull request by kyaroch**
Will resolve #27 

Screenshot_20190707_152619
Screenshot_20190707_142328
Screenshot_20190707_142352

### CODE REVIEW

File: lib/reconciliation.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
+require 'verbalize'
+
+class Reconciliation
+  include Verbalize::Action
+
+  input :lockbox_partner, :amount
+
+  def call
+    # The :amount input is the actual amount of money in the lockbox and must be
+    # a Money object
```
Comment (BintLopez) [inline lib/reconciliation.rb]: good comment, gracias!

File: lib/reconciliation.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+          raise ActiveRecord::Rollback
+        end
+      end
+      lockbox_action
+    end
+
+    result ? result : fail!(err_message)
+  end
+
+  def neutral_reconciliation
```
Comment (BintLopez) [inline lib/reconciliation.rb]: unused?

File: lib/reconciliation.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        raise ActiveRecord::Rollback
+      end
+
+      if amount != expected_amount
+        difference = expected_amount - amount
+        balance_effect = if difference.positive?
+          LockboxTransaction::CREDIT
+        else
+          LockboxTransaction::DEBIT
+        end
```
Comment (BintLopez) [inline lib/reconciliation.rb]: ^^ this little chunk (lines 27-32) could be a standalone private method

File: app/controllers/lockbox_partners/reconciliation_controller.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  def set_lockbox_partner
+    @lockbox_partner = LockboxPartner.find(params[:lockbox_partner_id])
+  end
+
+  def reconciliation_params
+    params.require(:reconciliation).permit(:amount)
+  end
+
+  def require_ownership
+    if !current_user.admin? && current_user.lockbox_partner != @lockbox_partner
```
Comment (BintLopez) [inline app/controllers/lockbox_partners/reconciliation_controller.rb]: We should have some controller tests around this access

File: Gemfile [hunk 5]
```diff
@@ -36,6 +36,8 @@ gem 'money-rails', '~>1.12'
 
 gem 'verbalize', '~> 2.2'
 
+gem 'material_icons', '~> 2.2'
```
Comment (BintLopez) [inline Gemfile]: let's use font awesome instead of material icons

File: app/controllers/lockbox_partners/reconciliation_controller.rb [hunk 6]
```diff
@@ -0,0 +1,37 @@
+require 'reconciliation'
+
+class LockboxPartners::ReconciliationController < ApplicationController
+  before_action :set_lockbox_partner, :require_ownership
+
+  def new
+  end
+
+  def create
+    amount = Monetize.parse(reconciliation_params[:amount])
```
Comment (BintLopez) [inline app/controllers/lockbox_partners/reconciliation_controller.rb]: reconciliation_params[:amount] is a float here, yeah?

File: app/models/lockbox_partner.rb [hunk 7]
```diff
@@ -2,6 +2,10 @@ class LockboxPartner < ApplicationRecord
   has_many :users
   has_many :lockbox_actions
 
+  # Number of days since last reconciliation when clinic user will be prompted
+  # to reconcile the lockbox. TODO make this configurable
```
Comment (BintLopez) [inline app/models/lockbox_partner.rb]: will you pretty please make an issue for this TODO?

File: lib/reconciliation.rb [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+class Reconciliation
+  include Verbalize::Action
+
+  input :lockbox_partner, :amount
+
+  def call
+    # The :amount input is the actual amount of money in the lockbox. We should
+    # handle conversion to Money outside of this class, to avoid making
+    # assumptions about the form
+    fail!("Amount is invalid") unless amount.is_a?(Money)
```
Comment (BintLopez) [inline lib/reconciliation.rb]: noice

File: app/controllers/lockbox_partners/reconciliation_controller.rb (hunk 6 above)
Comment (kyaroch) [reply app/controllers/lockbox_partners/reconciliation_controller.rb]: It's a string

File: spec/models/lockbox_partner_spec.rb [hunk 9]
```diff
@@ -257,4 +257,61 @@ def create_action_txn(partner, status)
...
+      end
+
+      context 'when the lockbox was created within the reconciliation interval' do
+        let(:created_at) { (LockboxPartner::RECONCILIATION_INTERVAL - 1).days.ago }
+
+        it { is_expected.to be false }
+      end
+
+      context 'when the lockbox was created outside the reconciliation interval' do
+        let(:created_at) { LockboxPartner::RECONCILIATION_INTERVAL.days.ago }
```
Comment (BintLopez) [inline spec/models/lockbox_partner_spec.rb]: does it make more sense for this to be off the eff_date of the cash added action vs the lockbox partner creation date? I'm imagining a scenario where we create a lockbox but due to onboarding and set up don't have partners using it immediately

File: app/models/lockbox_partner.rb [hunk 10]
```diff
@@ -36,4 +40,34 @@ def active?
...
+
+  def reconciliation_needed?
+    return false unless persisted?
+    return false unless !!reconciliation_interval_start
+    reconciliation_interval_start <= RECONCILIATION_INTERVAL.days.ago
+  end
+
+  def reconciliation_interval_start
+    # If the lockbox has never been reconciled, start counting from the date of
+    # the first cash addition
```
Comment (BintLopez) [inline app/models/lockbox_partner.rb]: good comment

**Pull request merged**